### ISSUE TITLE
PPD options from remotely shared printer unavailable

### ISSUE THREAD

**Issue by michaelweghorn** [apple/cups#5293]
With CUPS version 2.2.7, (PPD) options from printers shared by another host and set up using an URI like `ipp://<otherhost>/printers/<printername>` are no longer available.

Steps to reproduce this just using a single host:

Set up printer using a PPD file:

    $ sudo lpadmin -p myprinter -v ipp://printer1.example.org/ipp -P <SOME_PPD_FILE> -E

Set up another printer that refers to this printer (this would be done on another host usually):

    $ sudo lpadmin -p myprinter-from-remote-host -v ipp://localhost/printers/myprinter -E

Try to list available options:

    $ lpoptions -p myprinter-from-remote-host -l
    lpoptions: Unable to get PPD file for myprinter-from-remote-host: Not Found

This seems to have been introduced with commit 3645e5efa72a5773183949775bd73ebbcbbacbf1.
It works again as usual (options are listed) when I revert this commit on top of version 2.2.7 (I was using the version packaged in Debian as 2.2.7-2 for testing.)

In case this scenario is not supported, it would be good to know what the "correct" way to handle this use case would be (without using auto-discovery like DNS-SD etc.).

**Comment by michaelrsweet** [comment]
This was an intentional change to fix reported issues with this kind of remote printer setup. Basically, sandboxed applications on macOS and SELinux/AppArmor-protected applications on Linux don't always have remote network access, preventing these queues from working and causing some strange printing behavior.

The recommended configuration (since CUPS 1.4) has been to use a local PPD mirroring the remote printer capabilities - the simplest way to do this is to use the "everywhere" model (`lpadmin -p printer -E -v printer-uri -m everywhere`) which still sends the PostScript or PDF to the remote server unprocessed.

(FWIW, raw queues were officially deprecated in CUPS 2.2.7, but have been discouraged since CUPS 1.4...)

**Comment by michaelweghorn** [comment]
Thank you for your quick reply.

> The recommended configuration (since CUPS 1.4) has been to use a local PPD mirroring the remote printer capabilities - the simplest way to do this is to use the "everywhere" model (`lpadmin -p printer -E -v printer-uri -m everywhere`) which still sends the PostScript or PDF to the remote server unprocessed.

While this gives back *some* of the options provided in the PPD, the practical problem this puts us into is that important printer options are still missing, which unfortunately makes this a non-option for our current situation.

While CUPS is behaving "as designed" and all printers sold in the last years have *some* kind of support for IPP (and IPP Everywhere), we're still faced with the problem that relevant options are missing, which currently makes it impossible for us to switch away from deprecated APIs etc., s.a. #4969. (For quick testing, I used the PPD attached there and e.g. the PIN printing option was also missing when setting up the shared printer with the `-m everywhere` option.)

We'll get along with that for now (given e.g. that the above-mentioned commit has also been reverted for the CUPS package in Ubuntu 18.04), but it would be great if it were possible to have a better way of dealing with this in the future... (And while I really hope for that, I'm not 100 % sure that we can rely on all relevant devices properly supporting IPP (Everywhere) in the near future, at least in our setup....)
